src/server/services/wcs/qgswcsgetcapabilities.cpp

```diff
@@ -37,10 +37,53 @@ namespace QgsWcs
   void writeGetCapabilities( QgsServerInterface *serverIface, const QgsProject *project, const QString &version,
                              const QgsServerRequest &request, QgsServerResponse &response )
   {
-    QDomDocument doc = createGetCapabilitiesDocument( serverIface, project, version, request );
+    QStringList cacheKeyList;
+    bool cache = true;
 
-    response.setHeader( "Content-Type", "text/xml; charset=utf-8" );
-    response.write( doc.toByteArray() );
+    QgsAccessControl *accessControl = serverIface->accessControls();
+    if ( accessControl )
+      cache = accessControl->fillCacheKey( cacheKeyList );
+
+    QDomDocument doc;
+    QString cacheKey = cacheKeyList.join( '-' );
+    const QDomDocument *capabilitiesDocument = nullptr;
+
+    QgsServerCacheManager *cacheManager = serverIface->cacheManager();
+    if ( cacheManager && cache )
+    {
+      QByteArray content = cacheManager->getCachedDocument( project, request, cacheKey );
+      if ( !content.isEmpty() && doc.setContent( content ) )
+      {
+        doc = doc.cloneNode().toDocument();
+        capabilitiesDocument = &doc;
+      }
+    }
+
+    if ( !capabilitiesDocument ) //capabilities xml not in cache. Create a new one
+    {
+      doc = createGetCapabilitiesDocument( serverIface, project, version, request );
+
+      if ( cache && cacheManager )
+      {
+        if ( cacheManager->setCachedDocument( &doc, project, request, cacheKey ) )
+        {
+          QByteArray content = cacheManager->getCachedDocument( project, request, cacheKey );
+          if ( !content.isEmpty() && doc.setContent( content ) )
+          {
+            doc = doc.cloneNode().toDocument();
+            capabilitiesDocument = &doc;
+          }
+        }
+      }
+      if ( !capabilitiesDocument )
+      {
+        doc = doc.cloneNode().toDocument();
+        capabilitiesDocument = &doc;
+      }
+    }
+
+    response.setHeader( QStringLiteral( "Content-Type" ), QStringLiteral( "text/xml; charset=utf-8" ) );
+    response.write( capabilitiesDocument->toByteArray() );
```

Comment:
Warning, some naive questions are coming :).

Do you think that `QgsServerCacheManager` could manage some stuff like the `cacheKey` mechanism (`QgsAccessControl` could directly be used by the cache manager)?

Actually, do you think that we could replace:

```
    QStringList cacheKeyList;
    bool cache = true;

    QgsAccessControl *accessControl = serverIface->accessControls();
    if ( accessControl )
      cache = accessControl->fillCacheKey( cacheKeyList );
    QDomDocument doc;
    QString cacheKey = cacheKeyList.join( '-' );
    const QDomDocument *capabilitiesDocument = nullptr;

    QgsServerCacheManager *cacheManager = serverIface->cacheManager();
    if ( cacheManager && cache )
    {
      QByteArray content = cacheManager->getCachedDocument( project, request, cacheKey );
      if ( !content.isEmpty() && doc.setContent( content ) )
      {
        doc = doc.cloneNode().toDocument();
        capabilitiesDocument = &doc;
      }
    }
```

by something like:
```
QgsServerCacheManager *cacheManager = serverIface->cacheManager();  // QgsAccessControl accessed could be hide here
QDomDocument doc = cacheManager.getCachedDocument(project, request);  // hide QByteArray check, ... in this method

if ( ! doc )
{
   ...
}
```

Same question about the `setCachedDocument` action.